File: backend/app/schemas/play.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.sessions import SessionResponse
from app.schemas.conversation_snapshots import (
    ConversationSnapshotResponse,
    RestoreConversationSnapshotResponse,
)
from app.schemas.prompt_traces import (
    PromptTraceInspectorResponse,
    PromptTraceSummaryResponse,
)
from app.schemas.states import (
    SessionStateResponse,
    StateChangeLogResponse,
    StateSnapshotResponse,
)


class PlayOpeningOption(BaseModel):
    index: int
    title: str
    content: str
    is_default: bool = False


class PlayCardSummaryResponse(BaseModel):
    id: str
    name: str
    description: str | None = None
    tags: list[str] = Field(default_factory=list)
    cover_asset_id: str | None = None
    avatar_asset_id: str | None = None
    latest_session_id: str | None = None
    published_at: datetime | None = None
    opening_count: int = 0


class PlaySessionSummaryResponse(BaseModel):
    id: str
    name: str
    status: str
    card_id: str | None = None
    message_count: int = 0
    last_message_id: str | None = None
    last_message_at: datetime | None = None
    current_state_snapshot_id: str | None = None
    model_name: str | None = None
    created_at: datetime
    updated_at: datetime


class PlayCardDetailResponse(BaseModel):
    card: PlayCardSummaryResponse
    openings: list[PlayOpeningOption] = Field(default_factory=list)
    sessions: list[PlaySessionSummaryResponse] = Field(default_factory=list)


class PlaySessionOverviewResponse(BaseModel):
    session: SessionResponse
    card: PlayCardSummaryResponse
    openings: list[PlayOpeningOption] = Field(default_factory=list)
    state_summary: str = ""


class PlaySessionRenameRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)


class PlaySessionStatusRequest(BaseModel):
    status: str = Field(pattern="^(active|archived|deleted)$")


class PlayQuickReplyItem(BaseModel):
    id: str
    label: str
    content: str
    mode: str = "fill"
    order: int = 0


class PlayQuickReplyGroup(BaseModel):
    id: str
    name: str
    scope_type: str
    items: list[PlayQuickReplyItem] = Field(default_factory=list)


class PlaySessionExportResponse(BaseModel):
    session_id: str
    export_format: str
    export_scope: str
    file_name: str
    content: str


class PlaySnapshotListResponse(BaseModel):
    items: list[ConversationSnapshotResponse] = Field(default_factory=list)


class PlayTraceListResponse(BaseModel):
    items: list[PromptTraceSummaryResponse] = Field(default_factory=list)


class PlayStateBundleResponse(BaseModel):
    current: SessionStateResponse
    snapshots: list[StateSnapshotResponse] = Field(default_factory=list)
    changes: list[StateChangeLogResponse] = Field(default_factory=list)


class PlaySessionCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    opening_index: int | None = None
    model_name: str | None = None
    use_latest_existing_session: bool = False


class PlaySessionCopyRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    source_message_id: str | None = None


class PlaySessionCreateResponse(BaseModel):
    session: SessionResponse
    opening_message_id: str | None = None
    opening_selected: PlayOpeningOption | None = None

File: frontend/src/lib/api/play.ts
import { apiRequest } from "./client";

export type PlayOpeningOption = {
  index: number;
  title: string;
  content: string;
  is_default: boolean;
};

export type PlayCardSummary = {
  id: string;
  name: string;
  description: string | null;
  tags: string[];
  cover_asset_id: string | null;
  avatar_asset_id: string | null;
  latest_session_id: string | null;
  published_at: string | null;
  opening_count: number;
};

export type PlaySessionSummary = {
  id: string;
  name: string;
  status: string;
  card_id: string | null;
  message_count: number;
  last_message_id: string | null;
  last_message_at: string | null;
  current_state_snapshot_id: string | null;
  model_name: string | null;
  created_at: string;
  updated_at: string;
};

export type PlayCardDetail = {
  card: PlayCardSummary;
  openings: PlayOpeningOption[];
  sessions: PlaySessionSummary[];
};

export type SessionResponse = {
  id: string;
  mode: string;
  name: string;
  status: string;
  card_id: string | null;
  card_version_id: string | null;
  worldbook_id: string | null;
  project_id: string | null;
  persona_id: string | null;
  preset_version_id: string | null;
  origin_session_id: string | null;
  origin_snapshot_id: string | null;
  message_count: number;
  last_message_id: string | null;
  last_message_at: string | null;
  current_state_snapshot_id: string | null;
  model_name: string | null;
  created_at: string;
  updated_at: string;
};

export type PlaySessionCreateResponse = {
  session: SessionResponse;
  opening_message_id: string | null;
  opening_selected: PlayOpeningOption | null;
};

export type PlaySessionOverview = {
  session: SessionResponse;
  card: PlayCardSummary;
  openings: PlayOpeningOption[];
  state_summary: string;
};

export type CreatePlaySessionPayload = {
  name: string;
  opening_index?: number;
  model_name?: string;
  use_latest_existing_session?: boolean;
};

export function listPlayCards() {
  return apiRequest<PlayCardSummary[]>("play/cards");
}

export function getPlayCardDetail(cardId: string) {
  return apiRequest<PlayCardDetail>(`play/cards/${cardId}`);
}

export function createPlaySession(cardId: string, payload: CreatePlaySessionPayload) {
  return apiRequest<PlaySessionCreateResponse>(`play/cards/${cardId}/sessions`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function getPlaySessionOverview(sessionId: string) {
  return apiRequest<PlaySessionOverview>(`play/sessions/${sessionId}/overview`);
}
